src/els_diagnostics_server.erl

```diff
@@ -0,0 +1,168 @@
+%%==============================================================================
+%% Diagnostics Server
+%%==============================================================================
+-module(els_diagnostics_server).
+
+%%==============================================================================
+%% Behaviours
+%%==============================================================================
+-behaviour(gen_server).
+
+%%==============================================================================
+%% Exports
+%%==============================================================================
+
+-export([ start_link/2
+        ]).
+
+%% gen_server callbacks
+-export([ init/1
+        , handle_call/3
+        , handle_cast/2
+        , handle_info/2
+        ]).
+
+%% API
+-export([ on_save/1
+        , on_open/1
+        ]).
+
+%%==============================================================================
+%% Includes
+%%==============================================================================
+-include("erlang_ls.hrl").
+
+%%==============================================================================
+%% Macros
+%%==============================================================================
+
+%%==============================================================================
+%% Record Definitions
+%%==============================================================================
+-record(state, { server_name :: atom()
+               , uri :: uri()
+               , diagnostics :: term()
```

Comment:
It might make sense to keep the diagnostics in a map indexed by source (compiler, elvis, dialyzer) so that it is easy to add or update diagnostics for a new source.  This makes it easier to run the diagnostics sources async, and replace older diagnostics with new ones, but just for a given source.

Reply:
I like that idea 👍 I'd have to update the API to allow for that while currently we always run all diagnostics (as long as they're enabled) so I'd like to add the map but avoid adding the individual apis since they'd not be used yet. Wdyt?

Reply:
I am happy with small steps in the right direction.  I think we need to end up with a completely pluggable system, where we can load something with the appropriate behaviour, and register it as something contributing to the language server.